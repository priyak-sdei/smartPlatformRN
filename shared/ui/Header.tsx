import React, { ReactElement } from 'react';
import {
  View,
  TouchableOpacity,
  StyleSheet,
  ViewProps,
  ViewStyle,
} from 'react-native';
import Text from './Text';
import { fonts, moderateScale, spacing, useTheme } from '../theme';

interface HeaderProps extends ViewProps {
  title?: string;
  leftIcon?: ReactElement;
  leftIconColor?: string;
  leftTestID?: string;
  centerComponent?: ReactElement;
  rightComponent?: ReactElement;
  BottomBorder?: boolean;
  onLeftPress?: () => void;
  onRightPress?: () => void;
  style?: ViewStyle;
}

const Header: React.FC<HeaderProps> = ({
  title,
  leftIcon,
  leftTestID,
  centerComponent,
  rightComponent,
  BottomBorder,
  onLeftPress,
  onRightPress,
  style,
  ...rest
}) => {
  const { theme } = useTheme();
  const styles = getStyles(theme);
  return (
    <View
      style={[styles.container, BottomBorder && styles.bottomBorder, style]}
      {...rest}
    >
      <View style={styles.side}>
        {leftIcon ? (
          <TouchableOpacity
            onPress={onLeftPress}
            hitSlop={{ top: 10, bottom: 10, left: 10, right: 10 }}
            testID={leftTestID}
          >
            {leftIcon}
          </TouchableOpacity>
        ) : null}
      </View>
      <View style={styles.center}>
        {centerComponent
          ? centerComponent
          : !!title && (
            <Text style={styles.title} numberOfLines={1}>
              {title}
            </Text>
          )}
      </View>
      <View style={styles.side}>
        {rightComponent ? (
          <TouchableOpacity
            onPress={onRightPress}
            hitSlop={{ top: 10, bottom: 10, left: 10, right: 10 }}
          >
            {rightComponent}
          </TouchableOpacity>
        ) : null}
      </View>
    </View>
  );
};

const getStyles = (theme: any) =>
  StyleSheet.create({
    container: {
      flexDirection: 'row',
      height: moderateScale(50),
      backgroundColor: theme.colors.background,
      paddingHorizontal: moderateScale(spacing.s),
      justifyContent: 'space-between',
    },
    bottomBorder: {
      borderBottomWidth: StyleSheet.hairlineWidth,
      borderBottomColor: theme.colors.border,
    },
    side: {
      width: moderateScale(40),
      alignItems: 'center',
      justifyContent: 'center',
    },
    center: {
      flex: 1,
      alignItems: 'center',
      justifyContent: 'center',
      minWidth: 0,
    },
    title: {
      fontFamily: fonts.bold,
      fontSize: moderateScale(18),
      color: theme.colors.text,
      textAlign: 'center',
    },
  });

export default Header;
