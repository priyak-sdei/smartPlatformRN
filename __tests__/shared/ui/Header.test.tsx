import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import Header from '../../../shared/ui/Header';
import { Text as RNText } from 'react-native';
import renderer, { act } from 'react-test-renderer';

const mockTheme = {
    colors: {
        background: 'white',
        border: 'gray',
        text: 'black',
    },
};

jest.mock('../../../shared/theme', () => ({
    useTheme: () => ({ theme: mockTheme }),
    moderateScale: (v: number) => v,
    spacing: { s: 1, l: 2, xxs: 0.5 },
    fonts: { bold: 'bold' },
}));

jest.mock('../../../shared/ui/Text', () => {
    const { Text } = require('react-native');
    return (props: any) => <Text>{props.children || props.title}</Text>;
});

describe('Header', () => {
    it('renders with title', () => {
        const { getByText } = render(<Header title="Header Title" />);
        expect(getByText('Header Title')).toBeTruthy();
    });

    it('renders leftIcon and triggers onLeftPress', () => {
        const onLeftPress = jest.fn();
        const LeftIcon = () => <RNText>L</RNText>;
        const { getByText } = render(<Header leftIcon={<LeftIcon />} onLeftPress={onLeftPress} />);
        fireEvent.press(getByText('L'));
        expect(onLeftPress).toHaveBeenCalled();
    });

    it('renders rightComponent and triggers onRightPress', () => {
        const onRightPress = jest.fn();
        const RightIcon = () => <RNText>R</RNText>;
        const { getByText } = render(<Header rightComponent={<RightIcon />} onRightPress={onRightPress} />);
        fireEvent.press(getByText('R'));
        expect(onRightPress).toHaveBeenCalled();
    });

    it('renders centerComponent', () => {
        const Center = () => <RNText>CENTER</RNText>;
        const { getByText } = render(<Header centerComponent={<Center />} />);
        expect(getByText('CENTER')).toBeTruthy();
    });

    it('applies custom style', () => {
        const { getByTestId } = render(<Header title="Styled" style={{ backgroundColor: 'red' }} testID="header-root" />);
        expect(getByTestId('header-root')).toBeTruthy();
    });

    it('applies BottomBorder style', () => {
        const { getByTestId } = render(<Header title="Bordered" BottomBorder testID="header-root" />);
        expect(getByTestId('header-root')).toBeTruthy();
    });

    it('matches snapshot', () => {
        let tree: renderer.ReactTestRenderer | null = null;
        act(() => {
            tree = renderer.create(<Header title="Snapshot" />);
        });
        expect(tree!.toJSON()).toMatchSnapshot('Header component');
    });
}); 