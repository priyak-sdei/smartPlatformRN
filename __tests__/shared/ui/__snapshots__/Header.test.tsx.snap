// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header matches snapshot: Header component 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "white",
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "space-between",
        "paddingHorizontal": 1,
      },
      undefined,
      undefined,
    ]
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "justifyContent": "center",
        "width": 40,
      }
    }
  />
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "minWidth": 0,
      }
    }
  >
    <Text>
      Snapshot
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "justifyContent": "center",
        "width": 40,
      }
    }
  />
</View>
`;
